daemon/src/rollover_maker.rs

```diff
@@ -0,0 +1,296 @@
+            n_payouts,
+            oracle_pk,
+            sent_from_taker: None,
+            maker_cfd_actor: maker_cfd_actor.clone_channel(),
+            oracle_actor: oracle_actor.clone_channel(),
+            on_stopping: vec![on_stopping0.clone_channel(), on_stopping1.clone_channel()],
+            projection_actor,
+            proposal,
+        }
+    }
+
+    async fn handle_rollover_msg(&mut self, msg: RollOverMsg) -> Result<()> {
+        let sender = self
+            .sent_from_taker
+            .as_mut()
+            .context("cannot forward message to rollover task")?;
+        sender.send(msg).await?;
+        Ok(())
+    }
+
+    async fn handle_accept_rollover(
+        &mut self,
+        _msg: AcceptRollOver,
+        ctx: &mut xtra::Context<Self>,
+    ) -> Result<()> {
+        let order_id = self.cfd.order.id;
+
+        let (sender, receiver) = mpsc::unbounded();
+
+        self.sent_from_taker = Some(sender);
+
+        tracing::debug!(%order_id, "Maker accepts a roll_over proposal" );
+
+        let cfd = self.cfd.clone();
+
+        let dlc = cfd.open_dlc().expect("CFD was in wrong state");
+
+        let oracle_event_id = oracle::next_announcement_after(
+            time::OffsetDateTime::now_utc() + cfd.order.settlement_interval,
+        )?;
+
+        let taker_id = self.taker_id;
+
+        self.send_to_taker_actor
+            .send(maker_inc_connections::TakerMessage {
+                taker_id,
+                msg: wire::MakerToTaker::ConfirmRollOver {
+                    order_id,
+                    oracle_event_id,
+                },
+            })
+            .await??;
+
+        self.projection_actor
+            .send(UpdateRollOverProposal {
+                order: order_id,
+                proposal: None,
+            })
+            .await?;
+
+        let announcement = self
+            .oracle_actor
+            .send(oracle::GetAnnouncement(oracle_event_id))
+            .await?
+            .with_context(|| format!("Announcement {} not found", oracle_event_id))?;
+
+        let rollover = setup_contract::roll_over(
+            self.send_to_taker_actor.sink().with(move |msg| {
+                future::ok(maker_inc_connections::TakerMessage {
+                    taker_id,
+                    msg: wire::MakerToTaker::RollOverProtocol { order_id, msg },
+                })
+            }),
+            receiver,
+            (self.oracle_pk, announcement),
+            RolloverParams::new(
+                cfd.order.price,
+                cfd.quantity_usd,
+                cfd.order.leverage,
+                cfd.refund_timelock_in_blocks(),
+                cfd.order.fee_rate,
+            ),
+            Role::Maker,
+            dlc,
+            self.n_payouts,
+        )
+        .await;
+
+        match rollover {
+            Ok(dlc) => self.handle_updated_contract(order_id, dlc, ctx).await?,
+            Err(error) => self.handle_failed(order_id, error, ctx).await?,
+        };
```

Comment:
This won't work. You are blocking this handler function here until the protocol is complete, yet you expect it to make progress by taking messages out of `receiver` which is only populated by another handler function (line 278).

Reply:
I don't understand this. Sorry this might be a really dumb question.

Doesn't `setup_contract::roll_over().await` execute asynchronously, allowing other handler functions to run while it is executing?

the `sender` is set on line 157, `self.sent_from_taker = Some(sender)`,  before `setup_contract::roll_over().await`. The sender is used in `handle_rollover_msg` handler to send the `RolloverMsg` to the receiver. Shouldn't this happen concurrently?